Add default compression codec to MergeTree SETTINGS

gvoelfin opened the pull request:
Relates to: https://github.com/ClickHouse/ClickHouse/issues/42005

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new option to MergeTree SETTINGS which specifies a default compression codec in case the CREATE query does not explicitly define one for the given columns. This closes #42005.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
--> CREATE TABLE default_compression_codec_table (
    some_key UInt64 CODEC(DoubleDelta, LZ4), 
    some_id UInt32, 
    some_type String, 
    some_value Float64 CODEC(Gorilla)
) ENGINE = MergeTree()
ORDER BY (some_key)
SETTINGS default_compression_codec = 'LZ4';


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

gvoelfin committed:
Add default codec to MergeTree SETTINGS

Add a new option to MergeTree SETTINGS which specifies a default
copmression codec in case the CREATE query does not explicity defines
one for the given columns.

gvoelfin committed:
Remove unnused includes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=66394) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=66394) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=66394) it.</sub>

gvoelfin commented on src/Storages/ColumnsDescription.cpp:
```diff
@@ -869,6 +870,63 @@ bool ColumnsDescription::hasCompressionCodec(const String & column_name) const
+        auto [after_column_name, first] = std::invoke([this, &it]()->std::pair<String, bool>{
+            if(it == columns.begin())
+                return {"", true};
+            else
```
`columns_to_override` has just been created. `it` should never reach `columns.end()`

robot-ch-test-poll commented:
<!-- automatic status comment for PR #66394 from gvoelfin/ClickHouse:add_default_compression_codec_in_MergeTree_SETTINGS -->
*This is an automated comment for commit 1006f3dc4e6994f57388140ece7ccd76ff7c8d99 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66394/1006f3dc4e6994f57388140ece7ccd76ff7c8d99/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

antaljanosbenjamin commented on src/Storages/ColumnsDescription.cpp:
```diff
@@ -869,6 +870,63 @@ bool ColumnsDescription::hasCompressionCodec(const String & column_name) const
+        {
+            if(column.codec) continue;
+
+            String new_column_desc = std::invoke([&column, &codec_info_str]{
+                WriteBufferFromOwnString buf;
+                column.writeText(buf);
+                auto first_part = buf.str();
+                // Remove the unnecessary new line
+                first_part.pop_back();
+
+                WriteBufferFromOwnString whole_part;        
+                writeCString(first_part.c_str(), whole_part);
+                writeCString(codec_info_str.c_str(), whole_part);
+                return whole_part.str();
+            });
```
Here you assume that the codec is the last part for the column declaration, but that is [not true](https://clickhouse.com/docs/en/sql-reference/statements/create/table#with-explicit-schema). There can be a TTL expression, or even a [column level setting](https://clickhouse.com/docs/en/sql-reference/statements/create/table#with-explicit-schema) or statistics. Check the [implementation](https://github.com/ClickHouse/ClickHouse/blob/9c10d4f402b4c83c250a5bb4367aaa24142a21e6/src/Storages/ColumnsDescription.cpp#L145-L168) of `ColumnDescription::writeText`.

antaljanosbenjamin commented on src/Storages/ColumnsDescription.cpp:
```diff
@@ -869,6 +870,63 @@ bool ColumnsDescription::hasCompressionCodec(const String & column_name) const
     return it != columns.get<1>().end() && it->codec != nullptr;
 }
 
+void ColumnsDescription::setDefaultCompressionCodec(const String & codec_name)
```
In general I am not sure if we need to rewrite the query. `CompressionReadBufferBase` [seems to read the compression codec from the compressed buffer](https://github.com/ClickHouse/ClickHouse/blob/9c10d4f402b4c83c250a5bb4367aaa24142a21e6/src/Compression/CompressedReadBufferBase.cpp#L120), thus I don't think rewriting the create query is necessary.

I think in this way it can also be ambigouos in some cases. I will explain this in the main comment of PR.

antaljanosbenjamin reviewed (commented):
My main concern is why do we want to rewrite the query? In my mind this should work in a way like:
1. Check if the column has a codec, if yes, use that.
2. Check if the table had a default codec, if yes, use that.
3. Use the server-wide default codec.

Imagine the following usecase:

``` sql
CREATE TABLE x (a Int64)
ENGINE=MergeTree
ORDER BY x
SETTINGS default_compression_codec = 'Delta, ZSTD';

ALTER TABLE x MODIFY SETTINGS default_compression_codec = 'ZSTD';

-- I would expect that this insert uses the new default codec, but it wont. Only the newly created columns will use the new default codec
INSERT INTO x VALUES (1); 
```

UnamedRus commented:
> SETTINGS default_compression_codec = 'Delta, ZSTD';

It would be nice to have some tests with encrypting codecs as well, just in case. (It's not like they  are popular feature, but still)

gvoelfin committed:
Remove rewrite logic and add default codec

Revert the plan rewrite logic since it was not the right approach to
begin with. Add logic that checks if the default compression codec is
set in the mergetree settings before assigning it to the hardcoded
default one. Add integration tests to check if the used default encoding
is correct.

gvoelfin committed:
Reorder moved variables

gvoelfin committed:
Remove newline

gvoelfin commented:
@antaljanosbenjamin 
I think you are right, rewriting the query is not needed as you said, altering the table, changing the default compression codec would not yield the desired outcome. Removed the rewrite logic, and added the some checks into `MergeTreeDataPartWriter` and correctness is checked through checking the compression codec byte in the stored files. Added some cases for encrypted codecs as well. Will write the docs soon.

gvoelfin commented on tests/integration/test_default_compression_in_mergetree_settings/test.py:
```diff
@@ -0,0 +1,128 @@
+
+    node.query("INSERT INTO compression_table VALUES (1, 1, [[77]], 32)")
+
+    assert (
+        get_compression_codec_byte(
+            node, "compression_table", "1_0_0_0", "column_ok"
+        )
+        == CODECS_MAPPING["Multiple"]
+    )
+    assert (
+        get_compression_codec_byte(
+            node, "compression_table", "1_0_0_0", "column_ok.null"
+        )
+        == CODECS_MAPPING["LZ4"]
```
Not sure if I follow you here, if the column specific codec is removed from the table declaration, `column_ok` and `column_ok.null` both evaluate to `ZSTD` not `LZ4`. Should `LZ4` be the desired encoding here if the column specific encoding is not present? If that is the case I am not sure if I understand why.

antaljanosbenjamin replied:
Yes, you are right. However I have to check some details with some colleagues, please give me some time to make sure the changes are okay.

antaljanosbenjamin replied:
I checked out the stuff, see [this comment](https://github.com/ClickHouse/ClickHouse/pull/66394#discussion_r1681245740) of mine. It doesn't affect this PR though, I don't feel fair to ask you to make those changes. Also, don't stretch this PR more than necessary.

gvoelfin committed:
Extend docs

gvoelfin committed:
Style formatting

antaljanosbenjamin commented on docs/en/operations/settings/merge-tree-settings.md:
```diff
@@ -989,6 +989,12 @@ Used to regulate how resources are utilized and shared between mutations and oth
 
+## default_compression_codec
+
+Specifies the default compression codec to be used if none is defined for a particular column in the table declaration. If a column has a designated compression codec, it will override this default setting.
```
```suggestion
Specifies the default compression codec to be used if none is defined for a particular column in the table declaration. If a column has a designated compression codec, it will take priority over this default setting.
```

antaljanosbenjamin commented on src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp:
```diff
@@ -90,7 +90,14 @@ ASTPtr IMergeTreeDataPartWriter::getCodecDescOrDefault(const String & column_nam
+        if(!default_compression_codec_mergetree_settings.empty())
+            return CompressionCodecFactory::instance().get(default_compression_codec_mergetree_settings)->getFullCodecDesc();
+
+        return default_codec->getFullCodecDesc();
```
So a few things:
First of all, there are three very familiar functions in `StorageSnapshot`: https://github.com/ClickHouse/ClickHouse/blob/2315991504b1e95d7bb2594e54e3c6f749897d79/src/Storages/StorageSnapshot.cpp#L147C38-L186.
- `StorageSnapshot::getCodecDescOrDefault` is not used at all, it can be removed
- `StorageSnapshot::getCodecOrDefault` is only used in `LogSink` I think, so it can be moved into `LogSink`.

With these changes I would like to avoid the confusion of which function should be used when etc.


Second, this line `return default_codec->getFullCodecDesc();` shouldn't be executed at all. Therefore it makes sense to convert it to a `LOGICAL_ERROR`. So instead of returning, we should throw an exception with `LOGICAL_ERROR` with a message something about unexpected column name.

I don't think you have to do these changes, but wanted to document this somewhere public and this PR seemed the best place to do it.

gvoelfin committed:
Remove StorageSnapshot::getCodecDescOrDefault

gvoelfin committed:
Move StorageSnapshot::getCodecOrDefault into LogSink

gvoelfin committed:
Style format

gvoelfin committed:
Explicit lambda capture

gvoelfin committed:
Include MergeTreeSettings.h

gvoelfin committed:
Merge branch 'master' into add_default_compression_codec_in_MergeTree_SETTINGS

alexey-milovidov committed:
Update src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

alexey-milovidov committed:
Update docs/en/operations/settings/merge-tree-settings.md

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

tuanpach committed:
Merge remote-tracking branch 'ch/master' into add_default_compression_codec_in_MergeTree_SETTINGS

tuanpach committed:
Tidy some logic

tuanpach commented on src/Storages/StorageLog.cpp:
```diff
@@ -478,6 +482,32 @@ ISerialization::OutputStreamGetter LogSink::createStreamGetter(const NameAndType
+    if (const auto * virtual_desc = virtual_columns->tryGetDescription(column_name))
+        return get_codec_or_default(*virtual_desc);
+
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected column name: {}", column_name);
```
Hi @antaljanosbenjamin 

I throw an exception here if both `column_desc` and `virtual_desc` are nullptr instead of returning the default codec.
In `IMergeTreeDataPartWriter::getCodecDescOrDefault`, an exception is also thrown in this situation.

Do you think it is safe to throw an exception here after checking `column_desc` and `virtual_desc` ?

antaljanosbenjamin replied:
I think it should be fine. Hopefully if it is not fine, some tests will prove us wrong.

tuanpach committed:
Add default_compression_codec in MergeTreeSettingsChangesHistory

tuanpach committed:
Fix flaky test_default_compression_codec_in_mergetree_settings

tuanpach committed:
Fix test_default_compression_codec_in_mergetree_settings

tuanpach committed:
Update docs

tuanpach commented on src/Storages/MergeTree/MergeTreeSettings.cpp:
```diff
@@ -1771,6 +1771,14 @@ namespace ErrorCodes
     Calculate columns and secondary indices sizes lazily on first request instead
     of on table initialization.
     )", 0) \
+    DECLARE(String, default_compression_codec, "", R"(
```
Hi @antaljanosbenjamin 

I updated the doc for this setting. Please help to review it. Thanks

tuanpach committed:
Merge remote-tracking branch 'ch/master' into add_default_compression_codec_in_MergeTree_SETTINGS

tuanpach committed:
Trigger CI job

tuanpach committed:
Merge branch 'master' into add_default_compression_codec_in_MergeTree_SETTINGS

Blargian committed:
Correct path in feature_docs.py

antaljanosbenjamin commented on src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp:
```diff
@@ -1,9 +1,10 @@
-#include <Storages/MergeTree/IMergeTreeDataPartWriter.h>
-
 #include <Columns/ColumnSparse.h>
-#include <Common/MemoryTrackerBlockerInThread.h>
+#include <Compression/CompressionFactory.h>
+#include <Storages/MergeTree/IMergeTreeDataPartWriter.h>
 #include <Storages/MergeTree/MergeTreeIndexGranularity.h>
+#include <Storages/MergeTree/MergeTreeSettings.h>
 #include <Storages/StorageInMemoryMetadata.h>
+#include <Common/MemoryTrackerBlockerInThread.h>
```
I think the order should be main header first, and then the rest of the header.
```suggestion
#include <Storages/MergeTree/IMergeTreeDataPartWriter.h>

#include <Columns/ColumnSparse.h>
#include <Compression/CompressionFactory.h>
#include <Storages/MergeTree/MergeTreeIndexGranularity.h>
#include <Storages/MergeTree/MergeTreeSettings.h>
#include <Storages/StorageInMemoryMetadata.h>
#include <Common/MemoryTrackerBlockerInThread.h>
```

tuanpach replied:
Thanks.
I only enabled auto-format for modifications. So it didn't update the include part.

antaljanosbenjamin replied:
It is not that important, I won't hold back the PR because of this.

antaljanosbenjamin commented on src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp:
```diff
@@ -116,19 +122,21 @@ SerializationPtr IMergeTreeDataPartWriter::getSerialization(const String & colum
-    };
+    ASTPtr default_codec_desc = default_codec->getFullCodecDesc();
+
+    auto default_compression_codec_mergetree_settings = (*storage_settings)[MergeTreeSetting::default_compression_codec].toString();
```
Shouldn't this work? Then let's use to properly match the type.

```suggestion
    auto default_compression_codec_mergetree_settings = (*storage_settings)[MergeTreeSetting::default_compression_codec].value;
```

antaljanosbenjamin reviewed (commented):
Just two tiny changes.

tuanpach committed:
Update src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

tuanpach commented:
> Just two tiny changes.

Updated as per your suggestions.

tuanpach commented:
> @tuanpach Looks like the issue with the feature_check failing on docs not added is fixed now.

Many thanks.

tuanpach commented:
- [test_storage_rabbitmq](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=66394&sha=e6b288bc1bd974e5b23d20b71d8a29d983c285b1&name_0=PR&name_1=Integration%20tests%20%28tsan%2C%201%2F6%29): https://github.com/ClickHouse/ClickHouse/issues/71049
